ndcube Joss paper (#599)
sunpy/ndcube: joss_paper/paper.md

 - name: Aperio Software Ltd, UK
   index: 3
 - name: Meta Platforms Inc., UK
   index: 4
 - name: Workato Gmbh, Germany
   index: 5
 - name: Uptycs India Pvt. Ltd., India
   index: 6
 - name: Lockheed Martin Solar and Astrophysics Laboratory, USA
   index: 7
 - name: Bay Area Environmental Research Institute, USA
   index: 8
 - name: Claude Bernard Lyon 1 University, France
   index: 9
 - name: Institute of Theoretical Astrophysics, Norway
   index: 10
 - name: Space Telescope Science Institute, USA
   index: 11
 - name: NASA Goddard Space Flight Center, USA
   index: 12
 - name: European Space Agency, ESTEC
   index: 13
 - name: Southwest Research Institute, USA
   index: 14
date: 24 February 2023
bibliography: paper.bib

---

# Why Do We Need the ndcube Package?

N-dimensional data sets are common in all areas of science and beyond.
For example, a series of images taken sequentially with a CCD camera can be stored
as a single 3-D array with two spatial axes and one temporal axis.
In astronomy, the most commonly used framework for translating between array element
indices and the location, time, etc. in the Universe being observed is the World
Coordinate System (WCS).

> WCS’s ability to handle many different physical types (e.g. spatial, temporal,
> spectral, etc.) and projections (e.g. RA and Dec., helioprojective latitude and
> longitude, etc.) make it a succinct, standardized and powerful way to relate array
> axes to the physical types they represent.

Author: Good point.  I've just made a minor tweak to your suggestion here.

```suggestion
WCS’s ability to handle many different physical types of coordinates (e.g. spatial,
temporal, spectral, etc.) and their projections onto a data array (e.g. right ascension
and declination, helioprojective latitude and longitude, etc.) make it a succinct,
standardized and powerful way to relate array axes to the physical types they represent.
```

e.g. a sequence of 2-D spatial images arranged along a 3rd time axis.
In this case users can interact with the data as if it were a 3-D cube with a similar
API to `NDCube`.
Alternatively, the cubes can be ordered along one of the cubes' axes, e.g. a
sequence of tiles in an image mosaic where each cube represents an adjacent region
of the sky.
`NDCubeSequence` provides separate APIs for both the (N+1)-D and extended N-D
paradigms, enabling users to switch between them without reformatting the underlying
data.
`NDCubeSequence` also provides various methods to help with data analysis.
These APIs are similar to `NDCube` wherever possible, e.g. slicing, visualization,
etc., to minimize friction between analyzing single and multiple cubes.

`NDCollection` allows unordered but related `NDCube` and `NDCubeSequence` objects
to be linked, similar to how a Python dictionary is used.
However, in addition to dictionary-like features, `NDCollection` allows axes of
different cubes with the same lengths to be marked as 'aligned'.
This enables these axes on all constituent cubes to be sliced at the `NDCollection`
level.

> One application of this is linking derived data products, e.g. a spectral image cube
> and a Doppler may derived from one of its spectral lines.

Author: ```suggestion
One application of this is linking derived data products, e.g. a spectral image cube
and a Doppler map derived from one of its spectral lines.
```

By contrast, `GlobalCoords` supports scalar coordinates that apply to the whole
`NDCube` rather than any of its axes, e.g. the timestamp of a 2-D image.
WCS requires that all coordinates are associated with at least one array
axis, hence the need for `GlobalCoords`.
When an axis is dropped from an `NDCube` via slicing, the values of the dropped
coordinates at the relevant location along the dropped axis are automatically
added to the associated `GlobalCoords` object, e.g. the timestamp of a 2-D image
sliced from a space-space-time cube.
Thus coordinate information is never lost due to slicing.

`NDCube` objects are always instantiated with associated `ExtraCoords` and
`GlobalCoords` objects, even if empty.
Users can then add and remove coordinates subsequently.
For a more in depth dicussion of `ExtraCoords` and `GlobalCoords`, see @ndcube.

# Who Uses the ndcube Package?

The importance of the ndcube package is demonstrated by the fact that it already
supports a variety of current ground-based and satellite observatories.
These include the James Webb Space Telescope (JWST), Solar Orbiter,
the Interface Region Imaging Spectrograph (IRIS), Hinode, and the

> Daniel K. Inouye Solar Telescope (DKIST), as a dependency
> of the specutils [@specutils-docs; @specutils-code], sunraster [@sunraster],
> irispy-lmsal [@irispy-docs; @irispy-code], EISPAC [@eispac-docs; @eispac-code] and
> DKIST user tools packages, respectively.
> It is also used in the data pipeline of the PUNCH mission [Polarimeter to UNify

Author: ```suggestion
Daniel K. Inouye Solar Telescope (DKIST) via the
specutils [@specutils-docs; @specutils-code], jdaviz [@jdaviz], sunraster [@sunraster],
irispy-lmsal [@irispy-docs; @irispy-code], EISPAC [@eispac-docs; @eispac-code] and
DKIST user tools packages which all depend on ndcube.
ndcube is also used in the data pipeline of the PUNCH mission [Polarimeter to UNify
```

the Corona and Heliosphere; @punch], scheduled for launch in 2025.

> Individual researchers are also using the ndcube package in their own analysis

Author: ```suggestion
In addition, individual researchers are using the ndcube package in their own analysis
```
